src/Autofac/Core/Registration/ComponentRegistration.cs

```diff
@@ -202,6 +202,25 @@ public class ComponentRegistration : Disposable, IComponentRegistration
         /// <inheritdoc />
         public event EventHandler<IResolvePipelineBuilder>? PipelineBuilding;
 
+        /// <inheritdoc />
+        public void ConfigurePipeline(Action<IResolvePipelineBuilder> configurationAction)
```

Comment:
Would it be valuable to make this an extension method instead of something right on the interface? The only thing it actually needs that isn't already public is a way to determine if the pipeline has already been built; and given the very few times this is likely to be called, maybe that's just a `catch(InvalidOperationException)` instead of an `if (x is object)` check.

Alternatively, perhaps that informs the design of the interface in that there may need to be a better public way to check if the pipeline is built, so other folks _can_ write their own extension methods like this.

Reply:
Defining an extension method does make sense; also prevents the "technically" breaking change on `IComponentRegistration`.

The thing on my mind about exposing the pipeline state is that attempting to configure the pipeline for a registration after the pipeline has been built is by it's nature an exceptional action; it would only happen if the developer has made a mistake. 

Either the developer:
- Attaches to the right point, inside a module, before the container build has finished, so they're good, well-behaved and they get presents for Christmas, or...
- Naughty developers try to reach into the component registry while their code is running to try and bypass the immutable container, at which point we wag our finger at them and put coal in their stocking.

Basically, someone trying to write an extension that can react differently to a built or not-built pipeline is probably approaching things the wrong way.

Reply:
Okay, so trying to make it an extension method instead, but the problem is that if we rely on an exception being raised from the `ResolvePipeline` get, it means `ConfigurePipeline` is only valid to be called **if** ResolvePipeline throws an exception, so we'd be potentially getting a lot of exceptions raised.

I think this may have to stay on `ComponentRegistration`...while it is technically a breaking change, I can't imagine many people have created their own `IComponentRegistration` implementation.

Reply:
> it means `ConfigurePipeline` is only valid to be called if `ResolvePipeline` throws an exception

True, but that's also sort of the semantics around wrapping a `TryDoSomething` around a `DoSomething` method sometimes - the point might be to swallow the exception. Since `ConfigurePipeline` is only during registration time, is the throwing of the exceptions internally that big of a deal?

I'm not glued to the extension method thing, just thinking most of the time we've exposed enough on our interfaces for folks to write this sort of method themselves using extension methods; and all this does is add a convenience wrapper around an event that's already exposed. If someone _doesn't_ use `ConfigurePipeline` and instead goes directly to the `PipelineBuilding` event, there are other problems like "I added an event too late and it wasn't considered in the pipeline build."

Almost makes me wonder if the event handler for `PipelineBuilding` should be where the check for pipeline state is so if someone adds to the event too late they won't wonder why. There's no other way to really see if the pipeline has already been built.

Unless we add a convenience `PipelineState` property (`Built` or `Pending`?), which you'd need for attaching to the `PipelineBuilding` event anyway, at which point `ConfigurePipeline` could use that and it'd address the whole issue.

But like I said, I don't feel terribly strongly; if you do, say so and I'll just close this and call it a day.

Reply:
Okay; the fact that `ConfigurePipeline` throws, but `PipelineBuilding += {something}` does not feels wrong. So, modified to throw inside the `add` handler of `PipelineBuilding`, with an extension method to wrap it.